storage/key_index.go

```diff
@@ -0,0 +1,200 @@
+package storage
+
+import (
+	"bytes"
+	"errors"
+	"log"
+
+	"github.com/coreos/etcd/Godeps/_workspace/src/github.com/google/btree"
+)
+
+var (
+	ErrIndexNotFound = errors.New("index: not found")
+)
+
+// keyIndex stores the index of an key in the backend.
+// Each keyIndex has at least one key generation.
+// Each generation might have several key versions.
+// Tombstone on a key appends an tombstone version at the end
+// of the current generation and creates a new empty generation.
+// Each version of a key has an index pointing to the backend.
+//
+// For example: put(1);put(2);tombstone(3);put(4);tombstone(5) on key "foo"
+// generate a keyIndex:
+// key:     "foo"
+// index: 5
+// generations:
+//    {empty}
+//    {4, 5(t)}
+//    {1, 2, 3(t)}
+//
+// Compact a keyIndex removes the versions with smaller or equal to
+// index except the largest one. If the generations becomes empty
+// during compaction, it will be removed. if all the generations get
+// removed, the keyIndex Should be removed.
+
+// For example:
+// compact(2) on the previous example
+// generations:
+//    {empty}
+//    {4, 5(t)}
+//    {2, 3(t)}
+//
+// compact(4)
+// generations:
+//    {empty}
+//    {4, 5(t)}
+//
+// compact(5):
+// generations:
+//    {empty}
+//    {5(t)}
+//
+// compact(6):
+// generations:
+//    {empty} -> key SHOULD be removed.
+type keyIndex struct {
+	key         []byte
+	index       uint64
+	generations []generation
+}
+
+// put puts an index to the keyIndex.
+func (ki *keyIndex) put(index uint64) {
+	if index < ki.index {
```

Comment:
I think we could allow users set the same key twice in a multi-op(this gives the user more flexibility) and only records the last result in one `multi-op` because other results should not be seen at all.